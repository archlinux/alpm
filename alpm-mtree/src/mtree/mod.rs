//! Handling for the ALPM-MTREE file format.

pub mod v2;
use std::{
    fmt::{Display, Write},
    fs::File,
    io::{BufReader, Read},
    path::{Path, PathBuf},
    str::FromStr,
};

use alpm_common::{FileFormatSchema, MetadataFile};

use crate::{Error, MtreeSchema, mtree_buffer_to_string, parse_mtree_v2};

/// A representation of the [ALPM-MTREE] file format.
///
/// Tracks all available versions of the file format.
///
/// [ALPM-MTREE]: https://alpm.archlinux.page/specifications/ALPM-MTREE.5.html
#[derive(Clone, Debug, PartialEq, serde::Serialize)]
#[serde(untagged)]
pub enum Mtree {
    V1(Vec<crate::mtree::v2::Path>),
    V2(Vec<crate::mtree::v2::Path>),
}

impl MetadataFile<MtreeSchema> for Mtree {
    type Err = Error;

    /// Creates a [`Mtree`] from `file`, optionally validated using a [`MtreeSchema`].
    ///
    /// Opens the `file` and defers to [`Mtree::from_reader_with_schema`].
    ///
    /// # Note
    ///
    /// To automatically derive the [`MtreeSchema`], use [`Mtree::from_file`].
    ///
    /// # Examples
    ///
    /// ```
    /// use std::{fs::File, io::Write};
    ///
    /// use alpm_common::{FileFormatSchema, MetadataFile};
    /// use alpm_mtree::{Mtree, MtreeSchema};
    /// use alpm_types::{SchemaVersion, semver_version::Version};
    ///
    /// # fn main() -> testresult::TestResult {
    /// // Prepare a file with ALPM-MTREE data
    /// let file = {
    ///     let mtree_data = r#"#mtree
    /// /set mode=644 uid=0 gid=0 type=file
    /// ./some_file time=1700000000.0 size=1337 sha256digest=0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef
    /// ./some_link type=link link=some_file time=1700000000.0
    /// ./some_dir type=dir time=1700000000.0
    /// "#;
    ///     let mtree_file = tempfile::NamedTempFile::new()?;
    ///     let mut output = File::create(&mtree_file)?;
    ///     write!(output, "{}", mtree_data)?;
    ///     mtree_file
    /// };
    ///
    /// let mtree = Mtree::from_file_with_schema(
    ///     file.path().to_path_buf(),
    ///     Some(MtreeSchema::V2(SchemaVersion::new(Version::new(2, 0, 0)))),
    /// )?;
    /// # let mtree_version = match mtree {
    /// #     Mtree::V1(_) => "1",
    /// #     Mtree::V2(_) => "2",
    /// # };
    /// # assert_eq!("2", mtree_version);
    /// # Ok(())
    /// # }
    /// ```
    ///
    /// # Errors
    ///
    /// Returns an error if
    /// - the `file` cannot be opened for reading,
    /// - no variant of [`Mtree`] can be constructed from the contents of `file`,
    /// - or `schema` is [`Some`] and the [`MtreeSchema`] does not match the contents of `file`.
    fn from_file_with_schema(
        file: impl AsRef<Path>,
        schema: Option<MtreeSchema>,
    ) -> Result<Self, Error> {
        let file = file.as_ref();
        Self::from_reader_with_schema(
            File::open(file).map_err(|source| {
                Error::IoPath(PathBuf::from(file), "opening the file for reading", source)
            })?,
            schema,
        )
    }

    /// Creates a [`Mtree`] from a `reader`, optionally validated using a
    /// [`MtreeSchema`].
    ///
    /// Reads the `reader` to string (and decompresses potentially gzip compressed data on-the-fly).
    /// Then defers to [`Mtree::from_str_with_schema`].
    ///
    /// # Note
    ///
    /// To automatically derive the [`MtreeSchema`], use [`Mtree::from_reader`].
    ///
    /// # Examples
    ///
    /// ```
    /// use std::{fs::File, io::Write};
    ///
    /// use alpm_common::MetadataFile;
    /// use alpm_mtree::{Mtree, MtreeSchema};
    /// use alpm_types::{SchemaVersion, semver_version::Version};
    ///
    /// # fn main() -> testresult::TestResult {
    /// // Prepare a reader with ALPM-MTREE data
    /// let reader = {
    ///     let mtree_data = r#"#mtree
    /// /set mode=644 uid=0 gid=0 type=file
    /// ./some_file time=1700000000.0 size=1337 sha256digest=0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef
    /// ./some_link type=link link=some_file time=1700000000.0
    /// ./some_dir type=dir time=1700000000.0
    /// "#;
    ///     let mtree_file = tempfile::NamedTempFile::new()?;
    ///     let mut output = File::create(&mtree_file)?;
    ///     write!(output, "{}", mtree_data)?;
    ///     File::open(&mtree_file.path())?
    /// };
    ///
    /// let mtree = Mtree::from_reader_with_schema(
    ///     reader,
    ///     Some(MtreeSchema::V2(SchemaVersion::new(Version::new(2, 0, 0)))),
    /// )?;
    /// # let mtree_version = match mtree {
    /// #     Mtree::V1(_) => "1",
    /// #     Mtree::V2(_) => "2",
    /// # };
    /// # assert_eq!("2", mtree_version);
    /// # Ok(())
    /// # }
    /// ```
    ///
    /// # Errors
    ///
    /// Returns an error if
    /// - the `reader` cannot be read to string,
    /// - no variant of [`Mtree`] can be constructed from the contents of the `reader`,
    /// - or `schema` is [`Some`] and the [`MtreeSchema`] does not match the contents of the
    ///   `reader`.
    fn from_reader_with_schema(
        reader: impl std::io::Read,
        schema: Option<MtreeSchema>,
    ) -> Result<Self, Error> {
        let mut buffer = Vec::new();
        let mut buf_reader = BufReader::new(reader);
        buf_reader
            .read_to_end(&mut buffer)
            .map_err(|source| Error::Io("reading ALPM-MTREE data", source))?;
        Self::from_str_with_schema(&mtree_buffer_to_string(buffer)?, schema)
    }

    /// Creates a [`Mtree`] from string slice, optionally validated using a
    /// [`MtreeSchema`].
    ///
    /// If `schema` is [`None`] attempts to detect the [`MtreeSchema`] from `s`.
    /// Attempts to create a [`Mtree`] variant that corresponds to the [`MtreeSchema`].
    ///
    /// # Note
    ///
    /// To automatically derive the [`MtreeSchema`], use [`Mtree::from_str`].
    ///
    /// # Examples
    ///
    /// ```
    /// use std::{fs::File, io::Write};
    ///
    /// use alpm_common::MetadataFile;
    /// use alpm_mtree::{Mtree, MtreeSchema};
    /// use alpm_types::{SchemaVersion, semver_version::Version};
    ///
    /// # fn main() -> testresult::TestResult {
    /// let mtree_v2 = r#"
    /// #mtree
    /// /set mode=644 uid=0 gid=0 type=file
    /// ./some_file time=1700000000.0 size=1337 sha256digest=0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef
    /// ./some_link type=link link=some_file time=1700000000.0
    /// ./some_dir type=dir time=1700000000.0
    /// "#;
    /// let mtree = Mtree::from_str_with_schema(
    ///     mtree_v2,
    ///     Some(MtreeSchema::V2(SchemaVersion::new(Version::new(2, 0, 0)))),
    /// )?;
    /// # let mtree_version = match mtree {
    /// #     Mtree::V1(_) => "1",
    /// #     Mtree::V2(_) => "2",
    /// # };
    /// # assert_eq!("2", mtree_version);
    ///
    /// let mtree_v1 = r#"
    /// #mtree
    /// /set mode=644 uid=0 gid=0 type=file
    /// ./some_file time=1700000000.0 size=1337 sha256digest=0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef md5digest=d3b07384d113edec49eaa6238ad5ff00
    /// ./some_link type=link link=some_file time=1700000000.0
    /// ./some_dir type=dir time=1700000000.0
    /// "#;
    /// let mtree = Mtree::from_str_with_schema(
    ///     mtree_v1,
    ///     Some(MtreeSchema::V1(SchemaVersion::new(Version::new(1, 0, 0)))),
    /// )?;
    /// # let mtree_version = match mtree {
    /// #     Mtree::V1(_) => "1",
    /// #     Mtree::V2(_) => "2",
    /// # };
    /// # assert_eq!("1", mtree_version);
    /// # Ok(())
    /// # }
    /// ```
    ///
    /// # Errors
    ///
    /// Returns an error if
    /// - `schema` is [`Some`] and the specified variant of [`Mtree`] cannot be constructed from
    ///   `s`,
    /// - `schema` is [`None`] and
    ///   - a [`MtreeSchema`] cannot be derived from `s`,
    ///   - or the detected variant of [`Mtree`] cannot be constructed from `s`.
    fn from_str_with_schema(s: &str, schema: Option<MtreeSchema>) -> Result<Self, Error> {
        let schema = match schema {
            Some(schema) => schema,
            None => MtreeSchema::derive_from_str(s)?,
        };

        match schema {
            MtreeSchema::V1(_) => Ok(Mtree::V1(parse_mtree_v2(s.to_string())?)),
            MtreeSchema::V2(_) => Ok(Mtree::V2(parse_mtree_v2(s.to_string())?)),
        }
    }
}

impl Display for Mtree {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Self::V1(paths) | Self::V2(paths) => {
                    paths.iter().fold(String::new(), |mut output, path| {
                        let _ = write!(output, "{path:?}");
                        output
                    })
                }
            },
        )
    }
}

impl FromStr for Mtree {
    type Err = Error;

    /// Creates a [`Mtree`] from string slice `s`.
    ///
    /// Calls [`Mtree::from_str_with_schema`] with `schema` set to [`None`].
    ///
    /// # Errors
    ///
    /// Returns an error if
    /// - a [`MtreeSchema`] cannot be derived from `s`,
    /// - or the detected variant of [`Mtree`] cannot be constructed from `s`.
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Self::from_str_with_schema(s, None)
    }
}
