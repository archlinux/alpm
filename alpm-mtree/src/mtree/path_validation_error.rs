//! Error handling related to path validation based on [`Mtree`] data.

use std::{fmt::Display, path::PathBuf};

use alpm_types::Sha256Checksum;

#[cfg(doc)]
use crate::Mtree;

/// A list of errors that may occur when comparing [`Mtree`] data with paths inside a `base_dir`.
///
/// Tracks a `base_dir` whose files are compared to [`Mtree`] data.
/// Also tracks a list of zero or more [`PathValidationError`]s that occurred when validating paths
/// inside `base_dir` by comparing it with [`Mtree`] data.
///
/// After initialization, [`append`][`PathValidationErrors::append`] can be used to add any
/// errors to this struct that occurred during validation.
/// After validation (which is considered an infallible action), calling
/// [`check`][`PathValidationErrors::check`] returns an error if any errors have been collected
/// during validation.
#[derive(Debug, Default, thiserror::Error)]
pub struct PathValidationErrors {
    base_dir: PathBuf,
    errors: Vec<PathValidationError>,
}

impl PathValidationErrors {
    /// Creates a new [`PathValidationErrors`] for a directory.
    pub fn new(base_dir: PathBuf) -> Self {
        Self {
            base_dir,
            errors: Vec::new(),
        }
    }

    /// Appends a list of [`PathValidationError`]s to `self.errors`.
    pub fn append(&mut self, other: &mut Vec<PathValidationError>) {
        self.errors.append(other);
    }

    /// Checks if errors have been appended and consumes `self`.
    ///
    /// # Errors
    ///
    /// Returns an error if one or more errors have been appended.
    pub fn check(self) -> Result<(), crate::Error> {
        if !self.errors.is_empty() {
            return Err(crate::Error::PathValidation(self));
        }

        Ok(())
    }
}

impl Display for PathValidationErrors {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Errors occurred while comparing ALPM-MTREE data to paths in {:?}:\n{}",
            self.base_dir,
            self.errors.iter().fold(String::new(), |mut output, error| {
                output.push_str(&format!("{error}\n"));
                output
            })
        )
    }
}

/// The error that can occur when comparing [`Mtree`] paths with paths on a file system.
#[derive(Debug, thiserror::Error)]
pub enum PathValidationError {
    /// Alpm-common error.
    #[error("Alpm-common error:\n{0}")]
    AlpmCommon(#[from] alpm_common::Error),

    /// Unable to create hash digest for a path.
    #[error("Unable to create hash digest for path {path:?}:\n{source}")]
    CreateHashDigest {
        /// The path for which a hash digest can not be created.
        path: PathBuf,
        /// The source error.
        source: std::io::Error,
    },

    /// The hash digest of a path in the ALPM-MTREE data does not match that of the corresponding
    /// on-disk file.
    #[error(
        "The hash digest of {mtree_path:?} in the ALPM-MTREE data is {mtree_digest}, but that of {path:?} is {path_digest}"
    )]
    PathDigestMismatch {
        /// The path in the ALPM-MTREE data that does not have a matching path on disk.
        mtree_path: PathBuf,
        /// The size of the path according to ALPM-MTREE data.
        mtree_digest: Sha256Checksum,
        /// The on-disk path, that does not match the size of the ALPM-MTREE data.
        path: PathBuf,
        /// The on-disk path, that does not match the size of the ALPM-MTREE data.
        path_digest: Sha256Checksum,
    },

    /// The GID of a path in the ALPM-MTREE metadata does not match that of the corresponding
    /// on-disk file.
    #[error(
        "The GID of {mtree_path:?} in the ALPM-MTREE data is {mtree_gid}, but that of path {path:?} is {path_gid}"
    )]
    PathGidMismatch {
        /// The path in the ALPM-MTREE data that has a differing GID from that of the on-disk file.
        mtree_path: PathBuf,
        /// The GID recorded in the ALPM-MTREE data for `mtree_path`.
        mtree_gid: u32,
        /// The on-disk path, that has a differing GID from the ALPM-MTREE data.
        path: PathBuf,
        /// The GID used for `path`.
        path_gid: u32,
    },

    /// The metadata for a path can not be retrieved.
    #[error("The metadata for path {path:?} can not be retrieved:\n{source}")]
    PathMetadata {
        /// The on-disk path for which metadata can not be retrieved.
        path: PathBuf,
        /// The source Error.
        source: std::io::Error,
    },

    /// A path does not match what it is supposed to be.
    #[error("The path {mtree_path:?} in the ALPM-MTREE data does not match the path {path:?}")]
    PathMismatch {
        /// The path in the ALPM-MTREE data that does not have a matching path on disk.
        mtree_path: PathBuf,
        /// The on-disk path, that does not match that of the ALPM-MTREE data.
        path: PathBuf,
    },

    /// An on-disk path does not exist.
    #[error(
        "The path {path:?} does not exist, but the path {mtree_path:?} in the ALPM-MTREE data requires it to."
    )]
    PathMissing {
        /// The path in the ALPM-MTREE data that does not have a matching path on disk.
        mtree_path: PathBuf,
        /// The absolute on-disk path, that does not exist.
        path: PathBuf,
    },

    /// The mode of a path in the ALPM-MTREE metadata does not match that of the corresponding
    /// on-disk file.
    #[error(
        "The mode of {mtree_path:?} in the ALPM-MTREE data is {mtree_mode}, but that of path {path:?} is {path_mode}"
    )]
    PathModeMismatch {
        /// The path in the ALPM-MTREE data that has a differing mode from that of the on-disk
        /// file.
        mtree_path: PathBuf,
        /// The mode recorded in the ALPM-MTREE data for `mtree_path`.
        mtree_mode: String,
        /// The on-disk path, that has a differing mode from that of the ALPM-MTREE data.
        path: PathBuf,
        /// The mode used for `path`.
        path_mode: String,
    },

    /// An on-disk path is not a directory.
    #[error(
        "The path {path:?} is not a directory, but the ALPM-MTREE data for {mtree_path:?} requires it to be."
    )]
    PathNotADir {
        /// The path in the ALPM-MTREE data requiring `path` to be a directory.
        mtree_path: PathBuf,
        /// The absolute on-disk path, that is not a directory.
        path: PathBuf,
    },

    /// An on-disk path is not a file.
    #[error(
        "The path {path:?} is not a file, but the ALPM-MTREE data for {mtree_path:?} requires it to be."
    )]
    PathNotAFile {
        /// The path in the ALPM-MTREE data requiring `path` to be a file.
        mtree_path: PathBuf,
        /// The absolute on-disk path, that is not a file.
        path: PathBuf,
    },

    /// The size of a path in the ALPM-MTREE metadata does not match the size of the corresponding
    /// on-disk file.
    #[error(
        "The size of {mtree_path:?} in the ALPM-MTREE data is {mtree_size}, but that of path {path:?} is {path_size}"
    )]
    PathSizeMismatch {
        /// The path in the ALPM-MTREE data that does not have a matching path on disk.
        mtree_path: PathBuf,
        /// The size of the path according to ALPM-MTREE data.
        mtree_size: u64,
        /// The on-disk path, that does not match the size of the ALPM-MTREE data.
        path: PathBuf,
        /// The on-disk path, that does not match the size of the ALPM-MTREE data.
        path_size: u64,
    },

    /// A path does not match what it is supposed to be.
    #[error(
        "The symlink {mtree_path:?} in the ALPM-MTREE data points at {mtree_link_path:?}, while {path:?} points at {link_path:?}"
    )]
    PathSymlinkMismatch {
        /// The path in the ALPM-MTREE data that does not have a matching path on disk.
        mtree_path: PathBuf,
        /// The path in the ALPM-MTREE data that does not have a matching path on disk.
        mtree_link_path: PathBuf,
        /// The on-disk path, that does not match that of the ALPM-MTREE data.
        path: PathBuf,
        /// The on-disk path, that does not match that of the ALPM-MTREE data.
        link_path: PathBuf,
    },

    /// The time of a path in the ALPM-MTREE metadata does not match the time of the corresponding
    /// on-disk file.
    #[error(
        "The time of {mtree_path:?} in the ALPM-MTREE data is {mtree_time}, but that of path {path:?} is {path_time}"
    )]
    PathTimeMismatch {
        /// The path in the ALPM-MTREE data that does not have a matching path on disk.
        mtree_path: PathBuf,
        /// The size of the path according to ALPM-MTREE data.
        mtree_time: i64,
        /// The on-disk path, that does not match the size of the ALPM-MTREE data.
        path: PathBuf,
        /// The on-disk path, that does not match the size of the ALPM-MTREE data.
        path_time: i64,
    },

    /// The UID of a path in the ALPM-MTREE metadata does not match that of the corresponding
    /// on-disk file.
    #[error(
        "The UID of {mtree_path:?} in the ALPM-MTREE data is {mtree_uid}, but that of path {path:?} is {path_uid}"
    )]
    PathUidMismatch {
        /// The path in the ALPM-MTREE data that does not have a matching path on disk.
        mtree_path: PathBuf,
        /// The UID recorded in the ALPM-MTREE data.
        mtree_uid: u32,
        /// The on-disk path, that does not match the size of the ALPM-MTREE data.
        path: PathBuf,
        /// The UID used for `path`.
        path_uid: u32,
    },

    /// Unable to read a link.
    #[error(
        "The path {path:?} does not exist or is not a symlink, but the path {mtree_path:?} in the ALPM-MTREE data requires it to be:\n{source}."
    )]
    ReadLink {
        /// The path for a symlink on the file system which can not be read.
        path: PathBuf,
        /// The path in the ALPM-MTREE data that does not have a matching path on disk.
        mtree_path: PathBuf,
        /// The source error
        source: std::io::Error,
    },

    /// There are file system paths for which no matching ALPM-MTREE paths exist.
    #[error(
        "There are no matching ALPM-MTREE paths for the following file system paths:\n{}",
        paths.iter().fold(String::new(), |mut output, path| {
            output.push_str(&format!("{path:?}\n"));
            output})
    )]
    UnmatchedFileSystemPaths {
        /// The list of file system paths for which no matching ALPM-MTREE paths exist.
        paths: Vec<PathBuf>,
    },

    /// There are ALPM-MTREE paths for which no matching file system paths exist.
    #[error(
        "There are no matching file system paths for the following ALPM-MTREE paths:\n{}",
        paths.iter().fold(String::new(), |mut output, path| {
            output.push_str(&format!("{path:?}\n"));
            output})
    )]
    UnmatchedMtreePaths {
        /// The list of ALPM-MTREE paths for which no matching file system paths exist.
        paths: Vec<PathBuf>,
    },
}
