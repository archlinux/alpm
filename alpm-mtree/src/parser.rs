use std::path::PathBuf;

use alpm_parsers::iter_str_context;
use alpm_types::{Md5Checksum, Sha256Checksum};
use winnow::{
    ModalResult,
    Parser as WinnowParser,
    ascii::{digit1, line_ending, space0},
    combinator::{
        alt,
        cut_err,
        eof,
        fail,
        preceded,
        repeat_till,
        separated,
        separated_pair,
        terminated,
    },
    error::{StrContext, StrContextValue},
    stream::AsChar,
    token::{take_until, take_while},
};

use crate::path_decoder::decode_utf8_chars;

/// Each line represents a line in a .MTREE file.
#[derive(Debug, Clone)]
pub enum Statement<'a> {
    /// All lines that're irrelevant and don't contribute anything to the actual mtree file.
    ///
    /// Includes the following:
    /// - Empty lines
    /// - Lines that start with `#` (e.g. `#mtree` line and comments)
    Ignored,
    /// A `/set` command followed by some properties.
    Set(Vec<SetProperty<'a>>),
    /// A `/unset` command followed by some properties.
    Unset(Vec<UnsetProperty>),
    /// Any path statement followed by some properties.
    Path {
        path: PathBuf,
        properties: Vec<PathProperty<'a>>,
    },
}

/// Represents the properties that may be set in `/set` lines.
#[derive(Debug, Clone)]
pub enum SetProperty<'a> {
    Uid(u32),
    Gid(u32),
    Mode(&'a str),
    Type(PathType),
}

/// Represents the properties that can be unset by `/unset` lines.
#[derive(Debug, Clone)]
pub enum UnsetProperty {
    Uid,
    Gid,
    Mode,
    Type,
}

/// This type is used in a path line to define properties for that path.
#[derive(Debug, Clone)]
pub enum PathProperty<'a> {
    Uid(u32),
    Gid(u32),
    Mode(&'a str),
    Type(PathType),
    Size(u64),
    Link(PathBuf),
    Md5Digest(Md5Checksum),
    Sha256Digest(Sha256Checksum),
    Time(i64),
}

/// All allowed kinds of path types.
#[derive(Debug, Clone, Copy)]
pub enum PathType {
    Dir,
    File,
    Link,
}

/// Parse a single `/set` property.
fn set_property<'s>(input: &mut &'s str) -> ModalResult<SetProperty<'s>> {
    // First off, get the type of the property.
    let keywords = ["uid", "gid", "type", "mode"];
    let property_type = cut_err(alt(keywords))
        .context(StrContext::Label("property"))
        .context_with(iter_str_context!([keywords]))
        .parse_next(input)?;

    // Expect the `=` separator between the key-value pair
    let _ = "=".parse_next(input)?;

    // Now we continue parsing based on the type of the property.
    let property = match property_type {
        "type" => {
            let path_types = ["dir", "file", "link"];
            alt(path_types)
                .map(|value| match value {
                    "dir" => SetProperty::Type(PathType::Dir),
                    "file" => SetProperty::Type(PathType::File),
                    "link" => SetProperty::Type(PathType::Link),
                    _ => unreachable!(),
                })
                .context(StrContext::Label("property file type"))
                .context_with(iter_str_context!([path_types]))
                .parse_next(input)?
        }
        "uid" => SetProperty::Uid(system_id("user id", input)?),
        "gid" => SetProperty::Gid(system_id("group id", input)?),
        "mode" => SetProperty::Mode(mode(input)?),
        _ => unreachable!(),
    };

    Ok(property)
}

/// Parse a single `/unset` property.
fn unset_property(input: &mut &str) -> ModalResult<UnsetProperty> {
    // First off, get the type of the property.
    let keywords = ["uid", "gid", "type", "mode"];
    let property_type = cut_err(alt(keywords))
        .context(StrContext::Label("property"))
        .context_with(iter_str_context!([keywords]))
        .parse_next(input)?;

    // Map the parsed property type to the correct enum variant.
    let property = match property_type {
        "type" => UnsetProperty::Type,
        "uid" => UnsetProperty::Uid,
        "gid" => UnsetProperty::Gid,
        "mode" => UnsetProperty::Mode,
        _ => unreachable!(),
    };

    Ok(property)
}

/// Parse a simple system id as usize.
fn system_id(id_type: &'static str, input: &mut &str) -> ModalResult<u32> {
    cut_err(digit1.parse_to())
        .context(StrContext::Label(id_type))
        .context(StrContext::Expected(StrContextValue::Description(
            "a system id.",
        )))
        .parse_next(input)
}

/// Parse a Unix timestamp.
///
/// In mtree, this is a float for some reason, even though the decimal place is always a `0`.
fn timestamp(input: &mut &str) -> ModalResult<i64> {
    let (timestamp, _) = cut_err(separated_pair(digit1.parse_to(), '.', digit1))
        .context(StrContext::Label("unix epoch"))
        .context(StrContext::Expected(StrContextValue::Description(
            "A unix epoch in float notation.",
        )))
        .parse_next(input)?;

    Ok(timestamp)
}

/// Parse a filesystem mode.
///
/// Should be between 3-5 octal numbers **without** a `0o` prefix.
fn mode<'s>(input: &mut &'s str) -> ModalResult<&'s str> {
    cut_err(take_while(3..5, AsChar::is_oct_digit))
        .context(StrContext::Label("file mode"))
        .context(StrContext::Expected(StrContextValue::Description(
            "octal string of length 3-5.",
        )))
        .parse_next(input)
}

/// Parse a SHA-256 hash.
fn sha256(input: &mut &str) -> ModalResult<Sha256Checksum> {
    cut_err(take_while(64.., AsChar::is_hex_digit).parse_to())
        .context(StrContext::Label("sha256 hash"))
        .context(StrContext::Expected(StrContextValue::Description(
            "64 char long hexadecimal string",
        )))
        .parse_next(input)
}

/// Parse an MD5 hash.
fn md5(input: &mut &str) -> ModalResult<Md5Checksum> {
    cut_err(take_while(32.., AsChar::is_hex_digit).parse_to())
        .context(StrContext::Label("md5 hash"))
        .context(StrContext::Expected(StrContextValue::Description(
            "32 char long hexadecimal string",
        )))
        .parse_next(input)
}

/// Consume all chars of a link until a newline or space is hit.
///
/// Check [`decode_utf8_chars`] for more info on how special chars in paths are escaped.
fn link(input: &mut &str) -> ModalResult<String> {
    take_while(0.., |c| c != ' ' && c != '\n')
        .and_then(decode_utf8_chars)
        .parse_next(input)
}

/// Get a string representing a size by consuming all integers.
fn size(input: &mut &str) -> ModalResult<u64> {
    cut_err(take_while(0.., |c| c != ' ' && c != '\n').parse_to())
        .context(StrContext::Label("file size"))
        .context(StrContext::Expected(StrContextValue::Description(
            "a positive integer representing the file's size.",
        )))
        .parse_next(input)
}

/// Parse a single property.
fn property<'s>(input: &mut &'s str) -> ModalResult<PathProperty<'s>> {
    // First off, get the type of the property.
    let keywords = [
        "type",
        "uid",
        "gid",
        "mode",
        "size",
        "link",
        "md5digest",
        "sha256digest",
        "time",
    ];
    let property_type = cut_err(alt(keywords))
        .context(StrContext::Label("file property type"))
        .context_with(iter_str_context!([keywords]))
        .parse_next(input)?;

    // Expect the `=` separator between the key-value pair
    let _ = "=".parse_next(input)?;

    // Now we continue parsing based on the type of the property.
    let property = match property_type {
        "type" => alt(("dir", "file", "link"))
            .map(|value| match value {
                "dir" => PathProperty::Type(PathType::Dir),
                "file" => PathProperty::Type(PathType::File),
                "link" => PathProperty::Type(PathType::Link),
                _ => unreachable!(),
            })
            .context(StrContext::Label("property file type"))
            .context(StrContext::Expected(StrContextValue::Description(
                "'dir', 'file' or 'link'",
            )))
            .parse_next(input)?,
        "uid" => PathProperty::Uid(system_id("user id", input)?),
        "gid" => PathProperty::Gid(system_id("group id", input)?),
        "mode" => PathProperty::Mode(mode(input)?),
        "size" => PathProperty::Size(size.parse_next(input)?),
        "link" => PathProperty::Link(PathBuf::from(link.parse_next(input)?)),
        "md5digest" => PathProperty::Md5Digest(md5(input)?),
        "sha256digest" => PathProperty::Sha256Digest(sha256(input)?),
        "time" => PathProperty::Time(timestamp(input)?),
        _ => unreachable!(),
    };

    Ok(property)
}

/// Parse all path related properties that follow after a path declaration.
///
/// An example without all possible properties:
/// E.g. `./some_path uid=0 gid=0 type=file`
///                   ↑                   ↑
///                         This part
fn properties<'s>(input: &mut &'s str) -> ModalResult<Vec<PathProperty<'s>>> {
    cut_err(terminated(separated(0.., property, " "), line_ending)).parse_next(input)
}

/// Parse all properties that follow a `/set` command.
///
/// E.g. `/set uid=0 gid=0`
///            ↑         ↑
///             This part
fn set_properties<'s>(input: &mut &'s str) -> ModalResult<Vec<SetProperty<'s>>> {
    cut_err(terminated(separated(0.., set_property, " "), line_ending)).parse_next(input)
}

/// Parse all properties that follow an `/unset` command.
//////
/// E.g. `/unset uid gid`
///              ↑     ↑
///             This part
fn unset_properties(input: &mut &str) -> ModalResult<Vec<UnsetProperty>> {
    cut_err(terminated(separated(0.., unset_property, " "), line_ending)).parse_next(input)
}

/// Parse the next statement in the file.
fn statement<'s>(input: &mut &'s str) -> ModalResult<Statement<'s>> {
    // First, we figure out what kind of line we're looking at.
    let statement_type: String = alt((
        // A Path statement line
        //
        // Path statements may be preceded with whitespaces.
        // Otherwise read the line until terminated by the first space or newline.
        // Whitespace characters are encoded as `\s' (space), `\t' (tab), and `\n' (new line)
        // which is why we can simply ignore those while parsing the path.
        preceded(
            space0,
            terminated((".", take_until(0.., " ")).take(), alt((' ', '\n'))),
        ).and_then(decode_utf8_chars),
        terminated("/set", " ").map(|s: &str| s.to_string()),
        terminated("/unset", " ").map(|s: &str| s.to_string()),
        // A comment line that starts with `#`.
        preceded(("#", take_until(0.., "\n")), line_ending).map(|s: &str| s.to_string()),
        // An empty line that possibly contains spaces.
        preceded(space0, line_ending).map(|s: &str| s.to_string()),
        // If none of the above match, fail hard with a correct error message.
        fail.context(StrContext::Label("statement"))
        .context(StrContext::Expected(StrContextValue::Description(
            "'/set', '/unset', or a relative local path (./some/path) followed by their respective properties.",
        )))
    ))
    .parse_next(input)?;

    // Ignore comments and empty lines.
    if statement_type.trim().is_empty() {
        return Ok(Statement::Ignored);
    }

    // Now parse the properties based on the statement type until the end of line.
    let statement = match statement_type.as_str() {
        "/set" => Statement::Set(set_properties.parse_next(input)?),
        "/unset" => Statement::Unset(unset_properties.parse_next(input)?),
        path => Statement::Path {
            path: PathBuf::from(path),
            properties: properties.parse_next(input)?,
        },
    };

    Ok(statement)
}

/// Parse a given .MTREE file.
///
/// Empty lines and comment lines are returned as `Statement::Ignored`.
/// This is to provide a proper line-based representation of the file, so we can later on provide
/// proper context in error messages during the interpretation step.
///
/// # Errors
///
/// - `Error::ParseError` if a malformed MTREE file is encountered.
pub fn mtree<'s>(input: &mut &'s str) -> ModalResult<Vec<Statement<'s>>> {
    let (statements, _eof): (Vec<Statement<'s>>, _) =
        repeat_till(0.., statement, eof).parse_next(input)?;

    Ok(statements)
}
