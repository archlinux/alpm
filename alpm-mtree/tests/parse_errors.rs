//! Integration tests for the alpm-mtree parser.

use std::{fs::read_to_string, path::PathBuf};

use alpm_mtree::parser::mtree;
use insta::assert_snapshot;
use rstest::rstest;
use testresult::TestResult;
use winnow::Parser;

/// Mtree files are processed in two steps:
///
/// 1. Parsing the file. Make sure that the syntax is correct and return the syntax representation.
/// 2. Interpret the syntax. MTree files are stateful via the `/set` and `/unset` commands.
///
/// This test covers errors in the parsing step.
///
/// Test that faulty mtree files in the `parse_error_inputs` folder result in the expected error
/// messages. The error messages are stored in the `parse_error_snapshots` directory under the
/// respective name of the input file.
#[rstest]
fn ensure_errors_v1(#[files("tests/parse_error_inputs/*")] case: PathBuf) -> TestResult {
    // Read the input file and parse it.
    let input = read_to_string(&case)?;
    let result = mtree.parse(input.as_str());

    let Err(error) = result else {
        return Err(format!(
            "The parser succeeded even though it should've failed for input:\n{input}"
        )
        .into());
    };

    let name = case.file_stem().unwrap().to_str().unwrap();

    // Run the tests with the input being displayed as the description.
    // This makes reviewing this whole stuff a lot easier.
    // Also remove the usual module prefix by explicitly setting the snapshot path.
    // This isn't necessary, as we're already manually sorting snapshots by test scenario.
    let input_clone = input.clone();
    insta::with_settings!({
        description => input_clone,
        snapshot_path => "parse_error_snapshots",
        prepend_module_to_snapshot => false,
    }, {
        assert_snapshot!(name, error);
    });

    Ok(())
}
