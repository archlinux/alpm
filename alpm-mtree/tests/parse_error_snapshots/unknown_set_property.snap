---
source: alpm-mtree/tests/parse_errors.rs
description: "#mtree\n/set mode=755 id=0\n"
expression: error
---
parse error at line 2, column 15
  |
2 | /set mode=755 id=0
  |               ^
invalid property
expected `uid`, `gid`, `type`, `mode`
